contrib/load_key_clevis_hook.sh

```diff
@@ -0,0 +1,189 @@
+    fi
+  fi
+
+
+  CLEVIS_CHECK="$(zfs get -H -p -o value latchset.clevis:decrypt -s local $dataset_for_clevis_unlock)"
+  if [[ "$CLEVIS_CHECK" == "yes" ]]; then
+    zdebug "Found dataset for clevis unlocking: $dataset_for_clevis_unlock"
+    KEYLOCATION="$(get_fs_value "${dataset_for_clevis_unlock}" keylocation)"
+    if [[ -z "$(echo "$KEYLOCATION"|grep file)" ]]; then
+        # That's not us
+        zdebug "Warning: keylocation is not file while clevis unlock is set for dataset $dataset_for_clevis_unlock"
+        return 0
+    fi
+    # We suppose the keylocation has value in format file:///something/key
+    KEYLOCATION="${KEYLOCATION:7}"
+    KEYSTATUS="$(zfs get -H -p -o value keystatus -s none $dataset_for_clevis_unlock)"
+    if [[ "$KEYSTATUS" == "unavailable" ]]; then
+        # Prepare the key with password for unlocking in right place
+        mkdir -p "$(dirname "$KEYLOCATION")"
+        JWE="$(zfs get -H -p -o value latchset.clevis:jwe -s local $dataset_for_clevis_unlock)"
+        echo "$JWE" | clevis decrypt >"$KEYLOCATION"
+        cat "$KEYLOCATION" | zfs load-key -n $dataset_for_clevis_unlock
+        if [[ "$?" == "0" ]]; then
+          # We are fine
+          return 0
+        else
+          # We have autodecrypt flag set to on, however dataset can't be unlocked. Offer resealing the password
+          RESEAL="$(reseal_data_set "$dataset_for_clevis_unlock")"
+          if [[ "$RESEAL" != "" ]]; then
+            # We are fine
+            echo "$RESEAL"|clevis encrypt tpm2 '{"pcr_ids":"1,4,5,7,9","pcr_bank":"sha256"}' > /tmp/clevis_zfs.jwe
+            zdebug "Try to store correct jwe in $dataset_for_clevis_unlock"
+
+            # We need the pool to be writable
+            pool=${dataset_for_clevis_unlock%/*}
+            ro_state="$(zpool get -H -p -o value readonly "$pool")"
+            if [[ "$ro_state" = "on" ]]; then
+                zpool export "$pool"
+                zpool import -o readonly=off "$pool"
+                zfs set latchset.clevis:jwe="$(cat /tmp/clevis_zfs.jwe)" "$dataset_for_clevis_unlock"
+                zpool export "$pool"
+                zpool import -o readonly=on "$pool"
+            else
+                zfs set latchset.clevis:jwe="$(cat /tmp/clevis_zfs.jwe)" "$dataset_for_clevis_unlock"
+            fi
+
+            # check if we are fine
+            jwe_check="$(zfs get -H -p -o value latchset.clevis:jwe -s local $dataset_for_clevis_unlock)"
+            echo "$jwe_check" | clevis decrypt | zfs load-key -n -L prompt "$dataset_for_clevis_unlock"
+            if [[ "$?" != "0" ]]; then
+              # We failed. The jwe is not correctly stored in dataset
+              zdebug "Warning: We failed. The jwe is not correctly stored in dataset $dataset_for_clevis_unlock"
+              return 1
+            fi
+            echo "$RESEAL" >"$KEYLOCATION"
+            return 0
+          else
+            # We don't want to reseal
+            return 1
+          fi
+        fi
```

Comment:
I don't think any of this is a good idea. Handling these error conditions is best done interactively by a user who understands what's going on, not in an automatic hook.

Reply:
These are not error handlings. Those conditions are all valid. They just might happen. There is no need for the user to check where he/she is and what to do in this case. Resealing occures just once when ZBM, tpm2 and encrypted dataset are introduced to each other. After this introduction no resealing is required. The passphrase can be extracted from the hash stored in the special dataset property

Reply:
In that case, this definitely doesn't belong here. A one-shot initialization of the encryption scheme should happen long before ZFSBootMenu ever touches the pool.

Reply:
This is wrong. When zbm booter starts, it doesn't know how many times (if any) it started before on this host with this volume. There is no counter for that. So, we have to suppose, that this is not the first time as this is more likly. So we try to automatically unlock the dataset using this hook. And if during this unlocking process it turned out this is the first time we are running here, than only in this exclusive case we run resealing function. May be the name of this function is not completely relevant as it referes to something that happens quite often, while it is not true. It happens quite rare. Only when anyone of host hardware, ZBM or dataset has been changed since last boot. Suggest another name for the function, but it is sitting in a right place.

Reply:
I simply will not accept a script that operates as a load-key hook that does things like set properties on filesystems and, even worse, exports and re-imports the pool out from underneath the bootloader. It is potentially destructive, might unmount things that ZFSBootMenu has mounted and expected to be mounted, and doesn't even manage the import properly (all imports should use `-N` to suppress automount behavior).

Importing, mounting and exporting things in an `early_setup` hook are appropriate because ZFSBootMenu has not yet taken responsibility for any pools. At this stage in the execution, you should not be doing these manipulations. Whenever ZFSBootMenu is running, it should be considered the exclusive "owner" of the pool.

Further, as I noted before, these kinds of metadata setup operations are better suited to manual interactions with the user.

Your options at this point are to strip this logic from the script, or remove the script from this PR entirely so we can consider just the inclusion of hook support.

Reply:
Ok, I'll remove the script if it contradicts with the policy. I first tried early_setup hook, but it is inconvenient. I wrote about that in discussions section. BTW, may be you know a safe way to change user options in load_key hook? May be importing a pool with -N would be enough for that?

Reply:
I checked out several cases for importing/exporting pools using this hook with -N modifier for import. Everything works just fine. If you still insist on removing this hook from contrib, no problem, I'll do that. Just let me know your opinion.

Reply:
I decided to delete file for pr to concentrate on load_key hook. We can negotiate this hook inclusion in contrib as a separate pr later